apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -140,29 +85,37 @@ void ApplyEmulatorConfiguration()
         protected override GrpcAdapter DefaultGrpcAdapter =>
             throw new InvalidOperationException($"This property should never execute in {nameof(DatastoreDbBuilder)}");
 
-        private class EmulatorConfiguration
+        /// <summary>
+        /// Common code for handling project ID, namespace ID and emulator detection.
+        /// </summary>
+        private ConfiguredBuilder PrepareBuilder()
         {
-            internal const string EmulatorHostVariable = "DATASTORE_EMULATOR_HOST";
-            internal const string EmulatorProjectVariable = "DATASTORE_PROJECT_ID";
-            internal string Endpoint { get; }
-            internal string ProjectId { get; }
-
-            private EmulatorConfiguration(string endpoint, string projectId)
+            var emulatorBuilder = (DatastoreDbBuilder) MaybeUseEmulator();
+            if (emulatorBuilder is object)
             {
-                Endpoint = endpoint;
-                ProjectId = projectId;
+                return emulatorBuilder.PrepareBuilder();
```

Comment:
This is a recursive call?, it will work as long as EmulatorDetection defaults to None I think.